File: src/main/java/com/leet/medium/Triangle.go
package main

func minInt(a int, b int) int {
  if a <= b {
    return a
  }
  return b
}

func minimumTotal(triangle [][]int) int {
  prev := make([]int, len(triangle) + 1)
  for i := len(triangle) - 1; i >= 0; i-- {
    row := triangle[i]
    cur := make([]int, i + 1)
    for j, elem := range row {
      cur[j] = elem + minInt(prev[j], prev[j+1])
    }
    prev = cur
  }
  return prev[0]
}

File: src/main/java/com/leet/medium/Triangle.ts
/**
 * https://leetcode.com/problems/triangle/
 **/

function minimumTotal(triangle: number[][]): number {
  let paths = triangle[0];
  for (let i = 1; i < triangle.length; i++) {
    const size = triangle[i].length;
    // Next set of paths has size equal to current row
    const next: number[] = new Array(size);
    // Fill all the values, except first and last element
    for (let j = 1; j < size - 1; j++) {
      next[j] = triangle[i][j] + Math.min(paths[j], paths[j - 1]);
    }
    // Handle first and last values separately
    next[0] = triangle[i][0] + paths[0];
    next[size - 1] = triangle[i][size - 1] + paths[size - 2];
		// Update paths
    paths = next;
  }
	// Return minimum value
  return paths.reduce((acc, path) => Math.min(acc, path));
}
